### DISCUSSION TITLE
Host mailing preview on Next.js API route

### DISCUSSION THREAD

**Discussion by umar-ahmed** [sofn-xyz/mailing#345] [Ideas]
I see that Mailing.run is already built on top of Next.js. I was wondering if there is any interest in allowing projects to integrate the preview server in an existing Next.js project using API routes?

Example:

```js
// /pages/api/mailing/[...nextmailing].js
import NextMailing from 'mailing-next'

export default NextMailing({
  emailsDir: 'emails',
  preview: process.env.NODE_ENV === 'development'
})
```

The benefit of directly integrating is mostly for DX:

- Only need to spin up one server in local dev, ie. `next dev`
- Similar to the way [NextAuth](https://next-auth.js.org/getting-started/example#add-api-route) integrates
- Can share the same deployment, build tools, etc.

The main disadvantage I can see is that if Mailing.run decides to switch away for Next.js or not tie itself to any one framework, it might be harder to implement future features.

Let me know what you guys think!

**Comment by psugihara** [comment]
Hey @umar-ahmed I definitely like this idea as it might make it easier for some people to deploy. However, it seems kind of hard to implement. For example, I was not able to figure out a nice way to read the filesystem from a next API route so the next app actually runs [embedded in a vanilla node server](https://github.com/sofn-xyz/mailing/blob/main/packages/cli/src/commands/preview/server/start.ts#L47) when you're in development mode and that process packages up the emails directory to pass it to the next app in `.mailing`.

**Reply by umar-ahmed** [reply]
Hmm. Maybe an API route isn't the best place then. [next-sanity](https://github.com/sanity-io/next-sanity) actually uses a catch-all page instead.

I bet we could read the filesystem in getServerSideProps or getStaticProps instead. It might be a bit challenging to get live reload working though 🤔.

I'm planning on prototyping some of these ideas today. Let me know if you have any suggestions on where I could start. Ideally I'd like to re-use mailing-core, but will start from scratch for the preview server to make it easier to experiment

**Reply by psugihara** [reply]
Very cool! Yeah as a proof of concept, I would try just adding something (api or getServerSideProps or getStaticProps) to a nextjs app that reads the templates in emailsDir and renders them with mjml react (https://github.com/sofn-xyz/mailing/blob/main/packages/core/src/mjml.ts). You'd want it to work in both dev and production modes so that you could deploy your previews (`next build` may have been the blocker when I was trying to do this).

I think if you got that to work, the built in nextjs reload would just work for the hot reloading functionality (if not, we could get the existing websocket implementation running in an api route).


To explain the maybe relevant bits of mailing architecture... `packages/cli` is the CLI `mailing` app but also a next.js app. `npx mailling init` (or the alias `npx mailing`) copies most of CLI (the next.js app part) into `.mailing` and then runs `npx mailing preview` which runs that node server. The [node server](https://github.com/sofn-xyz/mailing/blob/main/packages/cli/src/commands/preview/server/start.ts#L70) proxies most requests to the next app in `.mailing` which it runs in development mode.

If you got the file reading stuff working in next.js, we could remove the vanilla nodejs server and proxy bit. Then I think it would be possible to write the `NextMailing` function you have above that would add all of the [pages](https://github.com/sofn-xyz/mailing/tree/main/packages/cli/src/pages) under a `/mailing` and `/api/mailing` namespace or something.

LMK if you have questions, and feel free to ping me in the discord.

**Reply by psugihara** [reply]
Just looked at sanity-next. Pulling the "routes" in with a wildcard route + routing component could work well too as long as the filesystem reading stuff works somehow.

**Reply by umar-ahmed** [reply]
Thanks for the info, that helps a lot!

I'll focus on figuring out that filesystem access question first then. If I figure anything out, I'll post an update.

**Reply by umar-ahmed** [reply]
@psugihara So I've got good news and I've got bad news:

## Good News

I was able to get an API route to read the emails directory. I tested it on the following environments:

- `next dev`
- `next build` and then `next start`
- Vercel (https://mailing-test-seven.vercel.app)

You can have a look at the two endpoints I implemented here:

- List templates: https://mailing-test-seven.vercel.app/api/mailing/list
- Preview template: https://mailing-test-seven.vercel.app/api/mailing/preview/AccountCreated

And you can see the full source code here: https://github.com/umar-ahmed/mailing-test

## Bad News

My current implementation relies on Webpack's `require.context()`. This works, and I'm able to read from the filesystem, but it comes with some limitations. The main one being that the parameters have to be static. Meaning, I can't load the `emailsDir` from a variable or config file. It has to be hard-coded to `"emails/"` like so: https://github.com/umar-ahmed/mailing-test/blob/3ddd420da9ed778365089e2e704c75a780af2e39/next-mailing/index.js#L7. I don't know if that is a deal-breaker or not?

Secondly, the live reload of Next.js doesn't seem to work out of the box with this. I feel like this is less of a problem since you mentioned that custom websockets idea, but I feel like it should be possible to make it work with Next.js' native live reloading since I am using Webpack's provided method to specify a folder from which I want to resolve modules. Webpack should be keeping track of that under the hood and marking files as dirty if they are changed in the filesystem. My hunch is that this just doesn't work with API routes since Next.js probably doesn't expect us to return html from it. I'm planning on trying a catch-all page instead to see if there's a way to get that to work, but I'll wait to hear back from you on that first limitation before I go any deeper

**Reply by psugihara** [reply]
Wow, nice work with `require.context()`! I didn't know you could do that. Custom `emailsDir` paths were an early feature we got a lot of requests for so I feel kind of conflicted on removing support. It's nice for users that have `src` directories in their next apps or are working in other frameworks with different structures.

Brainstorming...

- Would it work to hardcode several options for where the files might be? Then we could also check `src/emails` and perhaps other places that would make sense for remix/redwood apps. If we could do that I think it would be a reasonable trade-off to remove the option and just have a hardcoded list of spots we load templates from.

- I agree with your thoughts on the livereload piece. It might end up just not working though because of the non-standard import scheme. If it didn't, we'd want to make sure that we're able to watch the files. Here's the code for livereload if you want to do a simple test to make sure chokidar can watch those files and log changes: https://github.com/sofn-xyz/mailing/blob/main/packages/cli/src/commands/preview/server/livereload.ts#L13

- Do tsx templates work in a ts next app?

**Reply by umar-ahmed** [reply]
Okay, I played a bit with the code last night and I was able to incorporate the `emailsDir` using this code:

https://github.com/umar-ahmed/mailing-test/blob/76a9640b0a1fbcc540f69c51fa21bddf6a379e39/next-mailing/shared/cache.js#L8-L25

https://github.com/umar-ahmed/mailing-test/blob/76a9640b0a1fbcc540f69c51fa21bddf6a379e39/next-mailing/shared/context.js#L1-L5

This works in the sense that if I change the `emailsDir`, it will still import the email templates. However, because `require.context()` needs static parameters, right now this is hard-coded assuming a very specific path from the library to the project root directory (ie. `../../`). It basically scans for all the files in the project (excluding `node_modules`) and then filters that down based on the path.

The additional problem I see with this approach (besides having the path hard-coded) is that it will probably increase bundle size a lot because the require is so broad.

**Reply by umar-ahmed** [reply]
Bundle size actually is okay. Webpack is smart enough to extract out a separate chunk for this and it's only loaded server side, so it's not really an issue

**Reply by umar-ahmed** [reply]
Typescript files loading just fine:

<img width="432" alt="CleanShot 2022-12-04 at 17 16 53@2x" src="https://user-images.githubusercontent.com/8302959/205518610-1495a8ca-3ac1-4d5b-a4ad-b9a1ca4c4675.png">

**Reply by umar-ahmed** [reply]
Live reload:

https://user-images.githubusercontent.com/8302959/205518674-dc636d4d-62ee-4220-aa26-d7e4f537086c.mp4

**Reply by psugihara** [reply]
hell yes!

**Reply by umar-ahmed** [reply]
Ok I think I've found a decent compromise that removes the need to hard-code any paths. Basically, instead of using `require.context()` inside of the library code itself, we provide the user with a factory function that takes the `require.context()` has an argument:

```js
// emails/server.js
import { createNextMailing } from "next-mailing/server";

export const { handler, getServerSideProps } = createNextMailing(
  require.context(".", false, /\.[jt]sx$/)
);
```

and an equivalent for the page:

```js
// emails/client.js
import { buildNextMailingPage } from "next-mailing/client";

export const NextMailingPage = buildNextMailingPage();
```

These two files will need to be added to the boilerplate `emails` folder so that the user doesn't need to manually create them. Then the user can just add the pages to the correct routes:

```js
// pages/api/mailing/[[...nextmailing]].js
import { handler } from "emails/server";

export default handler;
```

```js
// pages/mailing/[[...nextmailing]].js
import { NextMailingPage } from "emails/client";
import { getServerSideProps } from "emails/server";

export default NextMailingPage;

export { getServerSideProps };
```

The client and server have to be separated like this so that Next.js correctly separates the dependencies (there are some calls to `node:fs` in mailing-core).

**Reply by psugihara** [reply]
love that solution for the emailsDir

Just looked at the code. Very clean, very promising!

I can explore how we might fold some of this into packages/cli and get rid of the custom node server more tomorrow. My first thought is that we could add the rest of the preview server features/ui to this then use it in the `.mailing` next app. We might want to do it in its own package (e.g. `packages/next-mailing`) so that we can configure the build separately from the CLI. Then when you're installing in a next app, we'd try to install in your app directly rather than generating the `.mailing` app.

**Reply by umar-ahmed** [reply]
Sounds good to me! 

You probably have a better idea than I do of how exactly we'd go about integrating this into the main repo, but what you've described makes sense to me. 

It's kind of like progressive enhancement; if the CLI detects that your project is Next.js project, it can integrate directly. And if not (say you're using Remix or something else), it'll just do what it currently does and create a separate folder. In that way, it's very similar to what Storybook has been doing with their CLI where it detects your framework and installs the appropriate packages and configs.

I'll leave https://github.com/umar-ahmed/mailing-test as is for your reference.

Also, if you start up a PR or want any help, feel free to ping me on GitHub or Discord. I'm happy to help 😄

**Reply by psugihara** [reply]
sounds good, appreciate this work so much 🙌. I'll definitely loop you for help as I try to integrate.

**Reply by psugihara** [reply]
Hey @umar-ahmed, I wanted to follow up here. I did some investigation and this seems doable but kind of hard to integrate (mostly a lot of routes to migrate). After reviewing priorities, unfortunately I don't think I'll have time in the near future.

If you wanted to take this work on yourself, I'd of course be happy to support. Otherwise we really appreciate the work and added a ticket with tasks for when someone does have time to pick it up here https://github.com/sofn-xyz/mailing/issues/402. Feel free to dm or email me if you want to chat more.
